service/controller/v26/resource/endpoint/current.go

```diff
@@ -113,6 +113,20 @@ func (r *Resource) GetCurrentState(ctx context.Context, obj interface{}) (interf
 				return nil, nil
 			}
 		}
+	} else if key.IsPodDeleted(pod) {
```

Comment:
So now we check for all pods if they still exist. The code right now treats masters and workers still differently though. For master pods we continue, for worker pods we cancel. Why is that? I would have expected to treat all pods the same way and either cancel all or continue all. Then the conditions in the code could be simplified as well. Otherwise I would appreciate more reasoning in the code comments so we understand this better still in the future.

Reply:
I didn't question the code above too much while writing this but it brings up some questions:
1. I would assume that if we cannot find the pod any more (it has been deleted) then the `endpoint` resource must have run successfully. Therefore removing the endpoint from the service already and making cancellation the correct choice. This would mean that this https://github.com/giantswarm/kvm-operator/blob/01f94ca5fa52bed212118c933abb8a5a09ac2712/service/controller/v26/resource/endpoint/current.go#L94-L99
should be a cancellation instead.

2. Master pods were treated differently so far because we actually waited for the containers inside the pod to be fully terminated before we removed the endpoint and finalizer. We did not do this for workers apparently. No clear reason why from my side, so we could align this if noone has an idea why we did this. The godoc on the condition also says:
```
	// Remove master endpoint ip only after pod is fully terminated because we
	// still need to drain the master before we remove it.
```
So no clue why we made a distinction between masters and workers here.

Are my assumptions correct?

Reply:
Uncertain, what I kind of realized now is that the endpoint resource does too much. It manages two different objectives which are about different lifecycles. One objective is the endpoint. Is it there or not. The lifecycle is then dictated by the cluster or eventually node pool. Then we have the other objective of a node or a node's IP to be managed. The lifecycle there is dictated by the nodes that exist in a cluster or eventually node pool. 

In any case, when something changes in this resource we should have tests that verify what we are changing. Just merging this and hoping for the best is shooting holes into the sky.

Reply:
I am fine with keeping this in a branch for now and figuring out if it helps by deploying it for a next update. 

The problem here has been that this issue has been around for a considerable amount of time and none of the people that tried to get closer to it were able to reproduce the behaviour we see in production. I had this idea and want to suggest it as one possible reason as to why this is happening (which would be an edge case). The issue seems generally rare and only happens only on large KVM clusters - like the ones we run on anubis.